### PR TITLE
fix(NODE-4932): remove .0 suffix from double extended json values

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

Removes the logic for integer-like doubles

#### What is the motivation for this change?

Fixes an output incompatibility with other languages

### Double check the following

- [ ] Ran `npm run lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by nbbeeken** [comment]
<details>
<summary>JShell tests</summary>

```java
import java.nio.ByteBuffer;

void assertEquals(String a, String b) throws Exception {
    if (!a.equals(b)) {
        throw new Exception(a + "does not equal" + b);
    }
}

char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();
String bytesToHex(byte[] bytes) {
    char[] hexChars = new char[bytes.length * 2];
    for (int j = 0; j < bytes.length; j++) {
        int v = bytes[j] & 0xFF;
        hexChars[j * 2] = HEX_ARRAY[v >>> 4];
        hexChars[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
    }
    return new String(hexChars);
}

assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("0"                       ) ).array() )   ,   "0000000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-0.0"                    ) ).array() )   ,   "8000000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-0.0"                    ) ).array() )   ,   "8000000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("3"                       ) ).array() )   ,   "4008000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-3"                      ) ).array() )   ,   "C008000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("3.4"                     ) ).array() )   ,   "400B333333333333")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("2.220446049250313e-16"   ) ).array() )   ,   "3CB0000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("123450000000"            ) ).array() )   ,   "423CBE3182800000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("1234.5"                  ) ).array() )   ,   "40934A0000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-1234.5"                 ) ).array() )   ,   "C0934A0000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("Infinity"                ) ).array() )   ,   "7FF0000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-Infinity"               ) ).array() )   ,   "FFF0000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("NaN"                     ) ).array() )   ,   "7FF8000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("1.7976931348623157e+308" ) ).array() )   ,   "7FEFFFFFFFFFFFFF")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("5e-324"                  ) ).array() )   ,   "0000000000000001")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-1.7976931348623157e+308") ).array() )   ,   "FFEFFFFFFFFFFFFF")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("-5e-324"                 ) ).array() )   ,   "8000000000000001")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("2.2250738585072014e-308" ) ).array() )   ,   "0010000000000000")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("2.225073858507201e-308"  ) ).array() )   ,   "000FFFFFFFFFFFFF")
assertEquals( bytesToHex( ByteBuffer.allocate(8).putDouble( Double.valueOf("5e-324"                  ) ).array() )   ,   "0000000000000001")
/exit
```
</details>

**Comment by nbbeeken** [comment]
[Evergreen patch](https://spruce.mongodb.com/version/63b863ce3627e05c84e249b4/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)

**Comment by nbbeeken** [comment]
[Evergreen patch](https://spruce.mongodb.com/version/63b86b190ae60643ac2c07eb/tasks)

**Comment by nbbeeken** [comment]
[Evergreen Patch](https://spruce.mongodb.com/version/63bc69ee3e8e863ed9aacc8d/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)

File: test/node/double_tests.js [hunk 1]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
+          // min positive subnormal number (NOTE: JS does not output same input string, but numeric values are equal)
+          input: '4.9406564584124654e-324',
+          output: { $numberDouble: '5e-324' }
+        },
+        {
+          // https://262.ecma-international.org/13.0/#sec-number.prototype.tofixed
+          // Note: calling toString on this integer returns 1000000000000000100, so toFixed is more precise
+          input: '1000000000000000128',
+          output: { $numberDouble: '1000000000000000128.0' }
+        }
```
Comment (baileympearson) [inline test/node/double_tests.js:85]: while this is interesting, I'm not sure this provides any additional value.  We already have tests that show that both positive and negative integer numbers serialize to numbers with a decimal.

File: test/node/double_tests.js [hunk 2]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
         const title = `returns ${inspect(output)} when Double is ${input}`;
         it(title, () => {
-          expect(output).to.deep.equal(input.toExtendedJSON({ relaxed: false }));
+          const inputAsDouble = new Double(input);
+          expect(inputAsDouble.toExtendedJSON({ relaxed: false })).to.deep.equal(output);
+          if (!Number.isNaN(inputAsDouble.value)) {
+            expect(Number(inputAsDouble.toExtendedJSON({ relaxed: false }).$numberDouble)).to.equal(
+              inputAsDouble.value
+            );
+          }
```
Comment (baileympearson) [inline test/node/double_tests.js:99]: I'm confused why the if-block and second assertion are necessary.  We should already be checking for numeric equality of `$numberDouble` in the preceding line (`expect(..).to.deep.equal(..)`).

File: test/node/double_tests.js [hunk 3]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
+        });
+
+        it(`input ${typeof input}: ${input} creates the same bytes after stringification`, () => {
```
Comment (baileympearson) [inline test/node/double_tests.js:102]: iiuc, the important aspect of this test is that the value of the double is preserved through ejson stringification.  we just happen to test this by asserting that the byte representations are equal.  Right?

If so, I propose we adjust the title to reflect what we're actually testing, and add a comment explaining that we determine that we're comparing bytes to assert that no loss of precision occurs.

```suggestion
        it(`preserves the precision of the number after stringification`, () => {
```

File: test/node/bson_corpus.spec.test.js [hunk 4]
```diff
@@ -181,22 +186,29 @@ describe('BSON Corpus', function () {
...
                 const eJSONParsed = EJSON.parse(cEJ, { relaxed: false });
                 expect(eJSONParsedAsJSON).to.have.nested.property('d.$numberDouble');
                 expect(eJSONParsed).to.have.nested.property('d._bsontype', 'Double');
                 const testInputAsFloat = Number.parseFloat(eJSONParsedAsJSON.d.$numberDouble);
+                const testInputAsNumber = Number(eJSONParsedAsJSON.d.$numberDouble);
                 const ejsonOutputAsFloat = eJSONParsed.d.valueOf();
-                expect(ejsonOutputAsFloat).to.equal(testInputAsFloat);
+                if (eJSONParsedAsJSON.d.$numberDouble === 'NaN') {
+                  expect(ejsonOutputAsFloat).to.be.NaN;
+                  expect(ejsonOutputAsFloat).to.be.NaN;
```
Comment (baileympearson) [inline test/node/bson_corpus.spec.test.js:204]: duplicate assertion?
```suggestion
                  expect(ejsonOutputAsFloat).to.be.NaN;
```

File: test/node/bson_corpus.spec.test.js [hunk 5]
```diff
@@ -220,18 +232,38 @@ describe('BSON Corpus', function () {
...
+                      : 'd.$numberLong'
+                    : 'd.$numberDouble'
+                );
                 expect(eJSONParsed).to.have.nested.property('d._bsontype', 'Double');
-                const testInputAsFloat = Number.parseFloat(eJSONFromBSONAsJSON.d.$numberLong);
                 const ejsonOutputAsFloat = eJSONParsed.d.valueOf();
-                expect(ejsonOutputAsFloat).to.equal(testInputAsFloat);
+                if (eJSONFromBSONAsJSON.d.$numberDouble === 'NaN') {
+                  expect(ejsonOutputAsFloat).to.be.NaN;
+                  expect(ejsonOutputAsFloat).to.be.NaN;
```
Comment (baileympearson) [inline test/node/bson_corpus.spec.test.js:259]:
```suggestion
                  expect(ejsonOutputAsFloat).to.be.NaN;
```

File: test/node/bson_corpus.spec.test.js [hunk 6]
```diff
@@ -165,6 +165,11 @@ describe('BSON Corpus', function () {
         describe('valid-extjson', function () {
```
Comment (baileympearson) [inline test/node/bson_corpus.spec.test.js:165]: it's unclear to me why the changes in this file are necessary.  reverting to what's currently in `main` doesn't make any tests fail.  could you explain why we're modifying the test runner?

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/node/bson_corpus.spec.test.js (hunk 4 above)
Comment (nbbeeken) [reply test/node/bson_corpus.spec.test.js]: Yep, fixed

File: test/node/bson_corpus.spec.test.js (hunk 5 above)
Comment (nbbeeken) [reply test/node/bson_corpus.spec.test.js]: fixed

File: test/node/double_tests.js (hunk 1 above)
Comment (nbbeeken) [reply test/node/double_tests.js]: This is an integer that's beyond the precision that an 8 byte float can hold (hence why toString loses the 28), it confirms that we still output a string that preserves it if that was the input. (Basically, asserts that toFixed is being used for ints).

File: test/node/double_tests.js (hunk 2 above)
Comment (nbbeeken) [reply test/node/double_tests.js:172]: Not exactly, the `$numberDouble` property is a string. This check converts it back to a number and checks that the string conversion did not loose the original value (within JavaScript).

File: test/node/double_tests.js (hunk 3 above)
Comment (nbbeeken) [reply test/node/double_tests.js]: Precision is related but more so that the string is acting as a correct data transfer format, we want to know that the same bytes can be reconstructed from the generated string, sometimes the string changes "4.9406564584124654e-324" -> "5e-324" but both represent the same ieee754 double bytes.

File: test/node/double_tests.js [hunk 7]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
-        { input: new Double(Number.EPSILON), output: { $numberDouble: '2.220446049250313e-16' } },
-        { input: new Double(12345e7), output: { $numberDouble: '123450000000.0' } },
-        { input: new Double(12345e-1), output: { $numberDouble: '1234.5' } },
-        { input: new Double(-12345e-1), output: { $numberDouble: '-1234.5' } },
-        { input: new Double(Infinity), output: { $numberDouble: 'Infinity' } },
-        { input: new Double(-Infinity), output: { $numberDouble: '-Infinity' } },
-        { input: new Double(NaN), output: { $numberDouble: 'NaN' } }
+        { input: 0, output: { $numberDouble: '0.0' } },
+        { input: -0, output: { $numberDouble: '-0.0' } },
+        { input: '-0.0', output: { $numberDouble: '-0.0' } },
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/node/double_tests.js]: Not true according to runtime, we do support string inputs, should we make that fix here or in another ticket?

As for the test I was just ensuring the decimal point (.0) doesn't break things here.

File: test/node/double_tests.js [hunk 8]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
+          // Note: calling toString on this integer returns 1000000000000000100, so toFixed is more precise
+          input: '1000000000000000128',
+          output: { $numberDouble: '1000000000000000128.0' }
+        }
       ];
 
       for (const test of tests) {
         const input = test.input;
         const output = test.output;
         const title = `returns ${inspect(output)} when Double is ${input}`;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/node/double_tests.js]: Sure, with what differentiating information?

File: test/node/bson_corpus.spec.test.js (hunk 6 above)
Comment (nbbeeken) [reply test/node/bson_corpus.spec.test.js]: With toFixed(1) reintroduced we don't need to skip the multi-type test anymore, I put that back, but it is senitive to changes in string format as opposed to value equivalence. 

The changes here are meant to raise issue in the future if there's a change to the output format that would break round tripping the string into a numeric value for all doubles. Since the if stmt on main filtered only for the `'1.2345678921232E+18'` case we did not need the extra handling for the NaN and -0 cases. The second section which tests round tripping through BSON needs to handle the cases where EJSON automatically converts values to int/long/double based on size and fraction.

File: test/node/double_tests.js [hunk 9]
```diff
@@ -38,26 +38,82 @@ describe('BSON Double Precision', function () {
...
-        { input: new Double(-3), output: { $numberDouble: '-3.0' } },
-        { input: new Double(3.4), output: { $numberDouble: '3.4' } },
-        { input: new Double(Number.EPSILON), output: { $numberDouble: '2.220446049250313e-16' } },
-        { input: new Double(12345e7), output: { $numberDouble: '123450000000.0' } },
-        { input: new Double(12345e-1), output: { $numberDouble: '1234.5' } },
-        { input: new Double(-12345e-1), output: { $numberDouble: '-1234.5' } },
-        { input: new Double(Infinity), output: { $numberDouble: 'Infinity' } },
-        { input: new Double(-Infinity), output: { $numberDouble: '-Infinity' } },
-        { input: new Double(NaN), output: { $numberDouble: 'NaN' } }
+        { input: 0, output: { $numberDouble: '0.0' } },
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/node/double_tests.js]: added titles

**Review by baileympearson — APPROVED** [review]
from discussion - we'll clean up the tests in a follow up PR but for the purposes of the fix we won't worry about test cleanup
